Failing test in Debian container

The CI tests on Debian currently fail due to a failing test (`FakeTempfileModuleTest.test_temporary_directory`):

``` text

_______________ FakeTempfileModuleTest.test_temporary_directory ________________

self = <pyfakefs.tests.fake_tempfile_test.FakeTempfileModuleTest testMethod=test_temporary_directory>

    def test_temporary_directory(self):
>       with tempfile.TemporaryDirectory() as tmpdir:
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

/work/pyfakefs/pyfakefs/tests/fake_tempfile_test.py:92: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3.13/tempfile.py:1141: in __exit__
    self.cleanup()
/usr/lib/python3.13/tempfile.py:1145: in cleanup
    self._rmtree(self.name, ignore_errors=self._ignore_cleanup_errors)
/usr/lib/python3.13/tempfile.py:1125: in _rmtree
    _rmtree(name, onexc=onexc)
/usr/lib/python3.13/shutil.py:763: in rmtree
    _rmtree_safe_fd(stack, onexc)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

stack = []
onexc = <function TemporaryDirectory._rmtree.<locals>.onexc at 0x7f014df67560>

    def _rmtree_safe_fd(stack, onexc):
        # Each stack item has four elements:
        # * func: The first operation to perform: os.lstat, os.close or os.rmdir.
        #   Walking a directory starts with an os.lstat() to detect symlinks; in
        #   this case, func is updated before subsequent operations and passed to
        #   onexc() if an error occurs.
        # * dirfd: Open file descriptor, or None if we're processing the top-level
        #   directory given to rmtree() and the user didn't supply dir_fd.
        # * path: Path of file to operate upon. This is passed to onexc() if an
        #   error occurs.
        # * orig_entry: os.DirEntry, or None if we're processing the top-level
        #   directory given to rmtree(). We used the cached stat() of the entry to
        #   save a call to os.lstat() when walking subdirectories.
        func, dirfd, path, orig_entry = stack.pop()
        name = path if orig_entry is None else orig_entry.name
        try:
            if func is os.close:
                os.close(dirfd)
                return
            if func is os.rmdir:
                os.rmdir(name, dir_fd=dirfd)
                return
    
            # Note: To guard against symlink races, we use the standard
            # lstat()/open()/fstat() trick.
>           assert func is os.lstat
                   ^^^^^^^^^^^^^^^^
E           AssertionError

/usr/lib/python3.13/shutil.py:663: AssertionError
```

Appearently, this is due to `shutils` patching not correctly working in this system under Python 3.13.5, probably related to the changes for #1171 in pyfakefs 5.9.0.
This does not happen under ubuntu, fedora and macOS with Python 3.13.7.

The call stack for `tempfile.py` seems not to match the lines in the respective Python release (the line numbers in the stack exceed the number of lines in the module), so this may also be due to some Debian-specific patches.

The installed Python version is listed as 3.13.5-1 (python3) and 3.13.5-2 (python3.13) - no idea why these seem to be different versions, but some patches may be involved.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -13,6 +13,11 @@ The released versions correspond to PyPI releases.
   the real filesystem behavior
 * remove support for Python versions before 3.10 (if needed, patches may be backported to the 5.x branch)
 
+### Unreleased
+
+## Fixes
+* fixes patching of Debian-specific `tempfile` in Python 3.13 (see [#1214](../../issues/1214))
+
 ## [Version 5.9.3](https://pypi.python.org/pypi/pyfakefs/5.9.3) (2025-08-28)
 Fixes a utility method.
 

--- a/pyfakefs/fake_filesystem_unittest.py
+++ b/pyfakefs/fake_filesystem_unittest.py
@@ -98,6 +98,7 @@ class TempfilePatcher:
 
     def __init__(self):
         self.tempfile_cleanup = None
+        self.tempfile_rmtree = None
 
     def start_patching(self):
         if self.tempfile_cleanup is not None:
@@ -109,6 +110,16 @@ class TempfilePatcher:
 
             self.tempfile_cleanup = tempfile._TemporaryFileCloser.cleanup  # type: ignore[module-attr]
             tempfile._TemporaryFileCloser.cleanup = cleanup  # type: ignore[module-attr]
+
+            if sys.version_info >= (3, 13) and hasattr(tempfile, "_rmtree"):
+                # Debian patches tempfile by importing or copying shutil.rmtree as _rmtree
+                # we patch this to use the original (patched) version
+                def _rmtree(*args, **kwargs):
+                    return tempfile._shutil.rmtree(*args, **kwargs)
+
+                self.tempfile_rmtree = tempfile._rmtree  # type: ignore[module-attr]
+                tempfile._rmtree = _rmtree  # type: ignore[module-attr]
+
         elif sys.platform != "win32":
 
             def close(self_, unlink=None):
@@ -125,6 +136,9 @@ class TempfilePatcher:
         else:
             tempfile._TemporaryFileCloser.cleanup = self.tempfile_cleanup  # type: ignore[module-attr]
         self.tempfile_cleanup = None
+        if self.tempfile_rmtree is not None:
+            tempfile._rmtree = self.tempfile_rmtree  # type: ignore[module-attr]
+            self.tempfile_rmtree = None
         # reset the cached tempdir in tempfile
         tempfile.tempdir = None
 
@@ -1063,12 +1077,13 @@ class Patcher:
             self._paused = False
 
             self.linecache_patcher.start_patching()
-            self.tempfile_patcher.start_patching()
 
             self.patch_modules()
             self.patch_functions()
             self.patch_defaults()
+
             self._set_glob_os_functions()
+            self.tempfile_patcher.start_patching()
 
             self._dyn_patcher = DynamicPatcher(self)
             sys.meta_path.insert(0, self._dyn_patcher)

--- a/pyfakefs/tests/fake_tempfile_test.py
+++ b/pyfakefs/tests/fake_tempfile_test.py
@@ -18,7 +18,6 @@ if using `Patcher` (via `fake_filesystem_unittest`).
 
 import os
 import stat
-import sys
 import tempfile
 import unittest
 
@@ -89,10 +88,6 @@ class FakeTempfileModuleTest(fake_filesystem_unittest.TestCase):
         self.assertTrue(self.fs.exists(dirname))
         self.assertEqual(self.fs.get_object(dirname).st_mode, stat.S_IFDIR | 0o700)
 
-    @unittest.skipIf(
-        sys.version_info[:3] == (3, 13, 5),
-        "Not working on Debian with this version, see #1214",
-    )
     def test_temporary_directory(self):
         with tempfile.TemporaryDirectory() as tmpdir:
             self.assertTrue(tmpdir)

